### ISSUE TITLE
Allow before/after TestMethod code with registering global callback

### ISSUE THREAD

**Issue by starosta33** [microsoft/testfx#2673]
the semantic should be something similar as below:

```
public class BeforeAfterTestMethodGlobalHook
{
    [GlobalBeforeTestMethod]
    public async Task BeforeTestMethodAsync(TestContext testContext)
    {
        // ...
    }

    [GlobalAfterTestMethod]
    public async Task AfterTestMethodAsync(TestContext testContext)
    {
        // ...
    }
}

**Comment by MarcoRossignoli** [comment]
@Evangelink I think we have already other request for this one.

**Comment by Evangelink** [comment]
Yes mine :)

**Comment by nohwnd** [comment]
What are the cases where you found it useful @evangelink?

If I am not mistaken this can be done with a base class, which has the flexibility that point 1. above seeks, and does not require duplicating of all the options (point 3.)

**Comment by Evangelink** [comment]
Good question, the base class is indeed the most forward solution.

I was mainly thinking about scenarios where you have already many tests and want to introduce a common behavior or when your tests already have base classes and you want an easier solution than creating X intermediate base class that would add this init/cleanup behavior and inherit from the previous base class.

**Comment by MarcoRossignoli** [comment]
I think that this feature is "useful" to move from other adapters to MSTest, this functionality is available there and looks like the update of big tests suite to achieve the same semantic is "too much".

**Comment by nohwnd** [comment]
> creating X intermediate base class that would add this init/cleanup behavior and inherit from the previous base class.

Another option here is to compose rather than inherit. So a helper that implements the setup would be added, and the existing base class would call it.

**Comment by Evangelink** [comment]
> Another option here is to compose rather than inherit. So a helper that implements the setup would be added, and the existing base class would call it.

Yes definitely! I was thiking about the cases where the base class is not something you own.

**Comment by nohwnd** [comment]
I did not think about that 👍

**Comment by Liamdoult** [comment]
I can share context of @starosta33's request. Currently our team uses NUnit and we cannot migrate to MSTtest until we have a solution.

In our test suite we are monitoring every remote call and performing some enforcement. These are our use-cases:
1. Ensure unit tests are clean and do not accidentally call remote dependencies.
1. Enforce a limited set of remote calls using an [AllowRemoteDependency("regex match uri")] attribute. We use this as both a security audit and to ensure our functional tests only hit the desired endpoints.
1. Explicit remote dependency checks. "Dependency x was called x times and returned 200 OK". This again is for security auditing, etc.
1. Disable all tests which use remote dependency X due to broken downstream dependency.

To do this we need to (at least in NUnit):
1. subscribe to events from HttpHandlerDiagnosticListener in some global onetime startup,
1. get the test context within the process so the event handler can match the event with the test, and
1. we need to perform some post test operation (in this case to assert the remote calls after the test is run).

This is used in every test throughout our test suite.

I believe if you use the inherit/compose approaches you are greatly limiting the extensibility of the library for some simple portable tooling. If we have to use inheritance/composition for every test class, this would require huge amounts of work (in our library ±1000 files) and would make the tool far less portable as every project will now have to update each of their classes.

But using the pipeline approach we can transparently add our behavior in and so can other users.
